File: _archive/api-py/app/api/v1/conversations.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from app.database import get_db
from app.api.deps import get_current_user
from app.utils.response import success_response, created_response
from app.models.conversation import Conversation, Message
from app.models.user import User

router = APIRouter(prefix='/conversations', tags=['Conversations'])


@router.get('/order/{order_id}')
async def get_conversation_by_order(order_id: str, db: AsyncSession = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await db.execute(select(Conversation).options(selectinload(Conversation.messages).selectinload(Message.sender)).where(Conversation.order_id == order_id))
    conv = result.scalar_one_or_none()
    if not conv:
        conv = Conversation(order_id=order_id)
        db.add(conv)
        await db.commit()
        await db.refresh(conv)
        return success_response({
            'conversation': {
                'id': conv.id,
                'orderId': conv.order_id,
                'messages': [],
                'unreadCount': 0,
            }
        })
    return success_response({
        'conversation': {
            'id': conv.id,
            'orderId': conv.order_id,
            'messages': [{'id': m.id, 'content': m.content, 'senderName': m.sender.name, 'senderRole': m.sender.role.value, 'isRead': m.read_at is not None, 'createdAt': m.created_at.isoformat()} for m in conv.messages],
            'unreadCount': sum(1 for m in conv.messages if m.read_at is None and m.sender_id != current_user.id),
        }
    })


@router.post('/order/{order_id}/message')
async def send_message(order_id: str, data: dict, db: AsyncSession = Depends(get_db), current_user: User = Depends(get_current_user)):
    result = await db.execute(select(Conversation).where(Conversation.order_id == order_id))
    conv = result.scalar_one_or_none()
    if not conv:
        conv = Conversation(order_id=order_id)
        db.add(conv)
        await db.flush()

    msg = Message(conversation_id=conv.id, sender_id=current_user.id, content=data.get('content'), type=data.get('type', 'TEXT'), media_url=data.get('media_url'))
    db.add(msg)
    await db.commit()
    await db.refresh(msg)
    return created_response({'message': {'id': msg.id, 'content': msg.content, 'createdAt': msg.created_at.isoformat()}}, 'Message sent')

File: apps/web/src/lib/api/conversations.ts
import { apiClient } from './client';

export interface Message {
  id: string;
  content: string;
  type: 'TEXT' | 'IMAGE' | 'FILE';
  mediaUrl?: string;
  senderId: string;
  senderName: string;
  senderRole: string;
  isRead: boolean;
  createdAt: string;
}

export interface Conversation {
  id: string;
  orderId: string;
  messages: Message[];
  unreadCount: number;
}

interface ConversationResponse {
  conversation: Conversation;
}

interface MessagesResponse {
  messages: Message[];
}

export const conversationsApi = {
  getByOrder: async (orderId: string): Promise<ConversationResponse> => {
    const conversation = await apiClient.get<Conversation>(`/conversations/order/${orderId}`);
    return { conversation };
  },

  sendMessage: async (orderId: string, data: { content: string; type?: string; mediaUrl?: string }): Promise<{ message: Message }> => {
    const message = await apiClient.post<Message>(`/conversations/order/${orderId}/message`, data);
    return { message };
  },

  getMessages: async (conversationId: string): Promise<MessagesResponse> => {
    return apiClient.get<MessagesResponse>(`/conversations/${conversationId}/messages`);
  },

  markAsRead: async (messageId: string): Promise<{ message: Message }> => {
    const message = await apiClient.patch<Message>(`/conversations/messages/${messageId}/read`);
    return { message };
  },
};
